src/shogun/base/progress.h

```diff
@@ -0,0 +1,199 @@
+        }
+
+        /** @class Wrapper for Range<T>::Iterator spawned by @ref PRange. */
+        class PIterator : public std::iterator<std::input_iterator_tag, T>
+        {
+            public:
+                PIterator(typename Range<T>::Iterator value, std::shared_ptr<ProgressPrinter> shrd_ptr)
+                : m_value(value), m_printer(shrd_ptr)
+                {
+                }
+                PIterator(const PIterator& other)
+                : m_value(other.m_value), m_printer(other.m_printer)
+                {
+                }
+                PIterator(PIterator&& other)
+                : m_value(other.m_value), m_printer(other.m_printer)
+                {
+                }
+                PIterator& operator=(const PIterator&) = delete;
+                PIterator& operator++()
+                {
+                    // Every time we update the iterator we print
+                    // also the updated progress bar
+                    m_printer->print_progress();
+                    m_value++;
+                    return *this;
+                }
+                PIterator& operator++(int)
+                {
+                    PIterator tmp (*this, this->m_printer);
+                    ++*this;
+                    return tmp;
+                }
+                T operator*()
+                {
+                    // Since PIterator is a wrapper we have
+                    // to return the actual value of the
+                    // wrapped iterator
+                    return *m_value;
+                }
+                bool operator!=(const PIterator& other)
+                {
+                    if (!(this->m_value != other.m_value))
+                        m_printer->print_end();
+                    return this->m_value != other.m_value;
+                }
+                bool operator==(const PIterator& other)
+                {
+                    return this->m_value == other.m_value;
+                }
+            private:
+                std::shared_ptr<ProgressPrinter> m_printer;
+                typename Range<T>::Iterator m_value;
+        };
+
+        /** Create the iterator that corresponds to the start of the range*/
+        PIterator begin() const {
+            return PIterator(m_range.begin(), m_printer);
+        }
+
+        /** Create the iterator that corresponds to the end of the iterator*/
+        PIterator end() const {
+            return PIterator(m_range.end(), m_printer);
+        }
+
+    private:
+        /** Range we iterate over */
+        Range<T> m_range;
+        /** Observer that will print the actual progress bar */
+        std::shared_ptr<ProgressPrinter> m_printer;
+    };
+
+    /** Creates @ref PRange given a range.
+     *
+     * @code
+     *  for (auto i : progress(range(0, 100), io)) { ... }
+     * @endcode
+     *
+     * @param   range   range used
+     */
+    template <typename T>
+    inline PRange<T> progress(Range<T> range, const SGIO & io)
```

Comment:
Let's add one more `progress` that uses global SGIO to simplify things. It's initialized with `init_shogun_with_defaults`